MIPs/mip-x .md

```diff
@@ -0,0 +1,81 @@
+---
+MIP: X
+Title: Fix Error Handling in API Responses
+Status: Draft
+Stability: Experimental
+discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/37
+Author(s): Shane Jonas (@shanejonas)
+Type: Maintaner
+Created: 2024-03-04
+---
+
+#### Abstract
+This proposal aims to refine MetaMask's API error handling by eliminating the practice of stringifying or wrapping errors. The current edge cases obscure underlying issues, complicating debugging and resolution for developers.
+
+#### Motivation
+Issue https://github.com/MetaMask/metamask-extension/issues/19697, https://github.com/MetaMask/metamask-extension/issues/15250 and https://github.com/MetaMask/metamask-extension/pull/15205#issuecomment-1199953855 highlighted difficulties faced by developers due to the API's error handling mechanism, where some errors are wrapped and stringified. This change proposes to break some existing error cases to facilitate clearer understanding and quicker resolution of issues.
+
+#### Specification
+- **Current Behaviour:** Errors returned by the API are wrapped and stringified, leading to terribly formatted error messages in some cases.
```

Comment:
Hmm, I had always thought of this as a bug, rather than a part of how the API works. But it's a fair point that we would expect dapps to expect these responses.

Do we know if we return these mangled stringified errors for all errors? Or just certain classes of them? I thought we returned proper Ethereum JSON-RPC errors at least some of the time.

Reply:
Would be good to drill in on what specific methods/scenarios we see these poorly formatted errors.